appmigrator_template.sh

```diff
@@ -0,0 +1,205 @@
+        finish_up "Current application is the same as the target."
+    fi
+
+    current_release_info=$(curl --silent --fail --retry 5 -X GET \
+    "${API_ENDPOINT}/v4/release?\$filter=belongs_to__application%20eq%20${APPLICATION_ID}" \
+    -H "Content-Type: application/json" \
+    -H "Authorization: Bearer $TOKEN") || finish_up "Couldn't get release data."
+    # If the user has no access to the app, the reply will be empty
+    local -a current_volumes=()
+    while IFS='' read -r volume; do
+        current_volumes+=("$volume")
+    done < <(echo "${current_release_info}" | jq -r '.d[0].composition.volumes | keys[]')
+    if [ ${#current_volumes[@]} -eq 0 ]; then
+        finish_up "Could not extract current volumes."
+    fi
+
+    echo "Current volumes:"
+    for v in "${current_volumes[@]}"; do
+        echo -n "$v"
+        if [ "${v}" == "${FROM_VOLUME:=resin-data}" ]; then
+            found_from_volume=true
+            echo " (selected)"
+        else
+            echo
+        fi
+    done
+    if [ "${found_from_volume}" != "true" ]; then
+        finish_up "Couldn't select volume to back up from."
+    fi
+
+    target_release_info=$(curl --silent --fail --retry 5 -X GET \
+    "${API_ENDPOINT}/v4/release?\$filter=belongs_to__application%20eq%20${TARGET_APP_ID}" \
+    -H "Content-Type: application/json" \
+    -H "Authorization: Bearer $TOKEN") || finish_up "Couldn't get release data."
+    # If the user has no access to the app, the reply will be empty
+    local -a target_volumes=()
+    while IFS='' read -r volume; do
+        target_volumes+=("$volume")
+    done < <(echo "${target_release_info}" | jq -r '.d[0].composition.volumes | keys[]')
+    if [ ${#target_volumes[@]} -eq 0 ]; then
+        finish_up "No volumes extracted to migrate to."
+    fi
+
+    if [ -z "${TO_VOLUME}" ] && [ ${#target_volumes[@]} -eq 1 ]; then
+        echo "Only one target volume, selecting that since no target volume was specified"
+        TO_VOLUME="${target_volumes[0]}"
+    fi
+    echo "Target app volumes:"
+    for v in "${target_volumes[@]}"; do
+        echo -n "$v"
+        if [ "${v}" == "${TO_VOLUME}" ]; then
+            found_to_volume=true
+            echo " (selected)"
+        else
+            echo
+        fi
+    done
+    if [ "${found_to_volume}" != "true" ]; then
+        finish_up "Couldn't select volume to back up to."
+    fi
+
+    echo "Stopping supervisor and user application"
+    systemctl stop resin-supervisor || true
+    if command balena >/dev/null 2>&1 ; then
+        ENGINE="balena"
+    else
+        ENGINE="docker"
+    fi
+    # shellcheck disable=SC2046
+    $ENGINE rm -f $($ENGINE ps -a -q) 2>/dev/null || true
+
+    # Pre-backup clean just in case
+    rm -rf "./backup" || finish_up "Couldn't clear temporary backup location."
+    mkdir -p ./backup || finish_up "Couldn't create temporary backup location."
+    rm -rf /mnt/data/backup.tgz || finish_up "Couldn't clear up backup file location."
```

Comment:
worth splitting out the backup location into a var (and maybe naming something more specific, for clarity?)

Reply:
Yeah, was thinking of that one, will update.